### PR TITLE
Fix rounding issue in `crop_success_probabilities()`

### PR DESCRIPTION
**Pull request by lloyddewit**
The `crop_success_probabilities()` endpoint should now give the same results as executing the function directly in R (including `prop_success` values after the decimal point).

For example, `crop_success_probabilities(country = "zm", station_id = "16", water_requirements = c(100, 300, 800))` should generate the following dataframe:

```
   station_name water_requirements planting_day planting_length prop_success
   <fct>                     <int>        <int>           <int>        <dbl>
 1 LUNDAZI MET                 100           92             120         0.33
 2 LUNDAZI MET                 100           92             180         0.33
 3 LUNDAZI MET                 100          122             120         0   
 4 LUNDAZI MET                 100          122             180         0   
 5 LUNDAZI MET                 100          153             120         0   
 6 LUNDAZI MET                 100          153             180         0   
 7 LUNDAZI MET                 300           92             120         0   
 8 LUNDAZI MET                 300           92             180         0   
 9 LUNDAZI MET                 300          122             120         0   
10 LUNDAZI MET                 300          122             180         0   
11 LUNDAZI MET                 300          153             120         0   
12 LUNDAZI MET                 300          153             180         0   
13 LUNDAZI MET                 800           92             120         0   
14 LUNDAZI MET                 800           92             180         0   
15 LUNDAZI MET                 800          122             120         0   
16 LUNDAZI MET                 800          122             180         0   
17 LUNDAZI MET                 800          153             120         0   
18 LUNDAZI MET                 800          153             180         0   
```

### CODE REVIEW

**Comment by lloyddewit** [comment]
@chrismclarke @istride 
In the `schema.py` files, it's possible to specify parameters as optional and provide a default value. While investigating the rounding issue, I came across a problem with these default values.

I originally specified default values that were sensible values for testing. However, I do not think this is correct. Instead, if the user does not specify a value when they call the endpoint, then we should use the parameter's default specified in the R function signature.

Some of the R function signatures specify `NULL` as the default value for some parameters. the Python equivalent of `NULL` is `None`. So I tried to set the default value of the API end points to `None`. Unfortunately however, Fast API does not seem to understand the concept of `None` and converts `None` values to a default value for that data type. 

For example, if the parameter is a list of integers, and I set the default value to `None`, then Fast API converts the `None` to a single-item list containing a zero (`[0]`). Similarly if the parameter is a boolean, then it converts `None` to `True`.

I worked around the list problem by setting the default to an empty list. Then in the function body, I check if the list is empty, and if so, then I set the parameter to `None`. The code below should make this clearer.

I don't yet have a solution for the boolean. I may need to create a new type that allows a boolean to have 3 values (`True`, `False` and `None`).

Does the above make sense?
Please could you review and sanity check my changes to the `schema.py` and `router.py` files? 
Do you have any ideas/suggestions about how best to handle `None` values in Fast API?

Happy to discuss if needed.
Thanks!

image

image

**Comment by lloyddewit** [comment]
@chrismclarke If you get a chance, please could you review this PR?
Thanks

**Comment by istride** [comment]
I did some experiments with FastAPI and found that it is only the API docs that show "[ 0 ]" as a default for a list of integers, in the examples. For that, it is possible to [declare request example data](https://fastapi.tiangolo.com/tutorial/schema-extra-example/).

I see that you've amended the code to use `List[int] | None`. This seems to be a Python >= 3.10 feature, so if we agree to only support those versions of Python then I believe we do not need to use `typing.List`, but can use simply `list` instead i.e. `list[int] | None`.

When it comes to booleans and default values, I cannot see a valid case for supporting three different values - surely, it has to be either true or false. In any case, I would consider `None` to be falsy.

**Comment by lloyddewit** [comment]
@istride Thank you so much for looking at this. A couple of hours after I wrote the comment, I realized what the problem was and resolved it. I updated the comment saying that it should be ignored. Sorry if this wasn't clear. In retrospect, I should have added an additional comment clarifying this.

However, your observations are still useful.

> I did some experiments with FastAPI and found that it is only the API docs that show "[ 0 ]" as a default for a list of integers, in the examples. For that, it is possible to [declare request example data](https://fastapi.tiangolo.com/tutorial/schema-extra-example/).

Yes correct, this is partly what was confusing me. Thank you also for the link, I wasn't aware of the possibility of declaring example data. This could be a useful feature.

> I see that you've amended the code to use `List[int] | None`. This seems to be a Python >= 3.10 feature, so if we agree to only support those versions of Python then I believe we do not need to use `typing.List`, but can use simply `list` instead i.e. `list[int] | None`.

Yes, I agree, I'll change the code to use `list`.

> When it comes to booleans and default values, I cannot see a valid case for supporting three different values - surely, it has to be either true or false. In any case, I would consider `None` to be falsy.

R allows a boolean to have 3 values (`TRUE`, `FALSE` or `NULL`). There are valid cases for this. For example, the R function we call has a boolean parameter. If the boolean is true or false, then the function uses this value directly. If the boolean is `NULL`, then the function sets the boolean to either true or false depending on the state of some other fields.

**Comment by chrismclarke** [comment]
Testing the default fastapi endpoint parameters locally I get the following error
```
  File "C:\Apps\e-picsa\epicsa-climate-api\.venv\Lib\site-packages\rpy2\rinterface.py", line 860, in __call__
    raise embedded.RRuntimeError(_rinterface._geterrmessage())
rpy2.rinterface_lib.embedded.RRuntimeError: Error in rpicsa::crops_definitions(data = daily, date_time = "date", station = "station",  :
  unused arguments (rain_totals = as.integer(water_requirements), plant_days = as.integer(planting_dates), plant_lengths = as.integer(crop_length))
```
Have double-checked SHA code on the epicsa-python lib as `c537501a76b666b8dc8df788970261029899db77` and that the R package install script passes successfully. Not sure if you get the same issue @lloyddewit or had similar in the past?

**Comment by lloyddewit** [comment]
@ChrisMarsh82 Thank you for testing.
I have a meeting so I have to rush ...
It looks like the correct R packages are still not getting installed by the `install_packages_picsa.R` script. We could try uninstalling the packages before we try and reinstall them. For example: `detach("package:epicsawrap", unload = TRUE)`.

**Comment by lloyddewit** [comment]
@chrismclarke Sorry, the message above was intended for you, not @ChrisMarsh82 (I selected the wrong auto-complete by mistake).

**Comment by lloyddewit** [comment]
@chrismclarke 
I updated `install_packages_picsa.R`.
Please could you run `Rscript install_packages_picsa.R` again and then retest?
Thanks

**Comment by chrismclarke** [comment]
After running the updated scripts still not working for me
```
  File "C:\Apps\e-picsa\epicsa-climate-api\.venv\Lib\site-packages\rpy2\rinterface.py", line 860, in __call__
    raise embedded.RRuntimeError(_rinterface._geterrmessage())
rpy2.rinterface_lib.embedded.RRuntimeError: Error in (function (country, station_id, planting_dates = NULL, water_requirements = NULL,  :
  unused argument (crop_length = list(0))
```
I even manually uninstalled everything using RStudio and ran the scripts again but still same error.
  
Have you been versioning the R packages? Here's a list of my installed packages

image

I also noticed the default parameters for the function have changed which might be part of the issue. Can you confirm the parameters you are using?

**Comment by chrismclarke** [comment]
I also noticed the default parameters for the function have changed which might be part of the issue. Can you confirm the parameters you are using?

Even when supplying better defaults I still get an issue 
```json
{
  "country": "zm",
  "station_id": "16",
  "water_requirements": [
    500
  ],
  "planting_length": [
    90
  ],
  "planting_dates": [
    120
  ],
  "start_before_season": true
}
```
```sh
File "C:\Apps\e-picsa\epicsa-climate-api\.venv\Lib\site-packages\rpy2\rinterface.py", line 860, in __call__
    raise embedded.RRuntimeError(_rinterface._geterrmessage())
rpy2.rinterface_lib.embedded.RRuntimeError: Error in (function (country, station_id, planting_dates = NULL, water_requirements = NULL,  :
  unused argument (crop_length = list(90))
  ```
  
Looking at the code it makes sense for there to be an error because the parameters do not expose `crop_length` parameter but instead `planting_length` (looks to be changed in db6d5f8b8ffa475e4ea0fea1d481e50cf357f8f6)

**Comment by lloyddewit** [comment]
@chrismclarke Sorry for the problems you're experiencing and thank you for providing details.

The new error message is different to the previous one. I believe you have the correct R packages installed. The error you are getting is because `epicsa-python` is trying to call `crop_success_probabilities()` with a `crop_length` parameter. This means that you have an older version of `epicsa-python` installed.

I checked the SHA in `requirements_picsa.txt` and it is correct.

Please run `pip install --force-reinstall -r requirements_picsa.txt`. If you still get the same error, then there's a problem with this instruction. Maybe we still need to specify `--no-cache-dir`, or if all else fails do a `pip uninstall epicsa-python`.
Hope this helps

**Comment by lloyddewit** [comment]
@chrismclarke I agree that we should use version numbers for the packages. I will add a version number to `epicsa-python`.

@lilyclements Please could you add version numbers to `rpicsa`, `epicsadata` and `epicsawrap`. 
I don't think we need a version number for `cdms.products` because I don't expect it to change and I still hope we can remove it.

**Comment by lloyddewit** [comment]
@chrismclarke 
Below are the function calls I use for testing. Each function should generate a `200` response (i.e. no errors).

```
annual_rainfall_summaries(country = "zm", station_id = "01122",
                                    summaries = c("annual_rain", "start_rains", "end_rains"))
annual_rainfall_summaries(country = "zm", station_id = "01122",
                                    summaries = c("annual_rain"))
annual_rainfall_summaries(country = "zm", station_id = "16", 
                                    summaries = c("start_rains", "end_rains", "annual_rain", "seasonal_rain"))
annual_rainfall_summaries(country = "zm", station_id = "16", 
                                    summaries = c("start_rains", "end_rains", "annual_rain", "seasonal_rain", "end_season"))

#epicsawrap issue #34
annual_rainfall_summaries(country = "zm", station_id = "1", 
                          summaries = c("end_season"))

annual_temperature_summaries(country = "zm", station_id = "16")
annual_temperature_summaries(country = "zm", station_id = "16", summaries = c("mean_tmin", "mean_tmax"))

crop_success_probabilities(country = "zm", station_id = "16")
crop_success_probabilities(country = "zm", station_id = "16", water_requirements = c(100, 300, 800))
crop_success_probabilities(country = "zm", station_id = "16", water_requirements = c(100, 300, 800), 
                           planting_length = c(100, 150), planting_dates = c(90, 100, 110))
crop_success_probabilities(country = "zm", station_id = "16", start_before_season = TRUE)
crop_success_probabilities(country = "zm", station_id = "16", start_before_season = FALSE)

monthly_temperature_summaries(country = "zm", station_id = "16")
monthly_temperature_summaries(country = "zm", station_id = "16", summaries = c("mean_tmin"))

season_start_probabilities(country = "zm", station_id = "16")
season_start_probabilities(country = "zm", station_id = "16", start_dates = c(10, 20, 100))
```

File: install_packages_picsa.R [hunk 1]
```diff
@@ -2,10 +2,25 @@
 # https://stackoverflow.com/questions/24646065/how-to-specify-lib-directory-when-installing-development-version-r-packages-from
 .libPaths()
 
-# Linked epicsa git repos
+# The `install_github()` lines below did not overwrite older installed package versions as expected.
```
[the remark this answers is not in the document]
Comment (lloyddewit) [reply install_packages_picsa.R:5]: The comment comes from my own observation of `install_github`'s behaviour. I could not find anything online about this.
Please see my comment below about version numbers.

**Comment by chrismclarke** [comment]
Bizarrely I tried the same reinstall script and this time it worked, so no idea what's going on with caching (if it's on the github side, python side or what)... But at least I'm seeing outputs now

Testing with 
```
{
  "country": "zm",
  "station_id": "1"
}
```
All probabilities are 0 which is a bit unexpected given that is one of the stations we are using for the workshop, but maybe a definition issue? I'll merge here and let others test on the live version

**Review by chrismclarke — APPROVED** [review]

**Pull request merged**